Fix flaky tests 2 (stateless, integration)

fm4v opened the pull request:
### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

Changes:
1. `ClickHouseCluster.get_docker_handle()` wait docker frequently and less time
2. integration tests: persist pdb history between test runs. 
Run pytest --pdb. On failure or breakpoint, it will fall back to PDB. Commands written here will be saved to a local file and loaded at the start of the next test run.
3. (attempt to) Fix some flaky integration and stateless tests   

List of tests:
- tests/integration/test_backup_restore_on_cluster/test.py::test_different_tables_on_nodes
- tests/integration/test_global_overcommit_tracker/test.py::test_global_overcommit
- tests/integration/test_merges_memory_limit/test.py::test_memory_limit_success
- tests/integration/test_runtime_configurable_cache_size/test.py::test_query_cache_size_is_runtime_configurable
- tests/integration/test_dictionaries_update_and_reload/test.py::test_reload_after_fail_by_timer
- 0_stateless/01287_max_execution_speed.sql
- 01676_clickhouse_client_autocomplete.sh
- 03002_filter_skip_virtual_columns_with_non_deterministic_functions.sql


Failed reports is [here](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAoY2hlY2tfbmFtZSBMSUtFICdJbnRlZ3JhdGlvbiUnIE9SIGNoZWNrX25hbWUgTElLRSAnU3RhdGVsZXNzJScpCiAgICBBTkQgY2hlY2tfc3RhcnRfdGltZSA+PSBub3coKSAtIElOVEVSVkFMIDYwIEhPVVIKICAgIEFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMAogICAgQU5EIHRlc3Rfc3RhdHVzICE9ICdTS0lQUEVEJwogICAgQU5EIHRlc3Rfc3RhdHVzIExJS0UgJ0YlJwogICAgQU5EIGNoZWNrX3N0YXR1cyAhPSAnc3VjY2VzcycKT1JERVIgQlkgY2hlY2tfbmFtZSwgdGVzdF9uYW1lLCBjaGVja19zdGFydF90aW1l)

fm4v committed:
integration tests: fix flaky tests

fm4v committed:
integration tests: `ClickHouseCluster.get_docker_handle()` wait docker frequently and less time

fm4v committed:
stateless tests: fix flaky tests

tavplubix commented on tests/integration/test_backup_restore_on_cluster/test.py:
```diff
@@ -164,8 +163,15 @@ def test_replicated_database():
     node2.query("INSERT INTO mydb.tbl VALUES (2, 'count')")
     node1.query("INSERT INTO mydb.tbl VALUES (3, 'your')")
     node2.query("INSERT INTO mydb.tbl VALUES (4, 'chickens')")
+    node1.query("OPTIMIZE TABLE mydb.tbl ON CLUSTER 'cluster' FINAL")
```
I guess it was trying to check backup of multiple parts, and now there's only one part

fm4v replied:
Then we need to compare parts before and after backup, not data. I will add separate test on it.

tavplubix commented on tests/integration/test_runtime_configurable_cache_size/test.py:
```diff
@@ -101,9 +101,11 @@ def test_query_cache_size_is_runtime_configurable(start_cluster):
     node.query("SELECT 2 SETTINGS use_query_cache = 1, query_cache_ttl = 1")
     node.query("SELECT 3 SETTINGS use_query_cache = 1, query_cache_ttl = 1")
 
+    time.sleep(2)
```
This is unreliable, it will remain flaky

fm4v replied:
I actually break this test [here](https://s3.amazonaws.com/clickhouse-test-reports/0/8c6a6f2b71efe8500efd94de1177372e35bf1ae5/integration_tests__release__[1_4].html) in the last assert, sleep is required as i understand
```python
        res = node.query_with_retry(
            "SELECT value FROM system.asynchronous_metrics WHERE metric = 'QueryCacheEntries'",
            check_callback=lambda result: result == "1\n",
        )
>       assert res == "1\n"
E       AssertionError: assert '2\n' == '1\n'
E         - 1
E         + 2

```

rschu1ze replied:
I wrote this test originally and found this open PR by luck.

There were three recorded test failures in the past 6 months or so:
- https://s3.amazonaws.com/clickhouse-test-reports/57933/ca3ca8ffe403a26a1ff5bc77adb8895c89b16970/integration_tests__tsan__[1_6].html
- https://s3.amazonaws.com/clickhouse-test-reports/60211/c5d3aef83e287b477f67bb5b0e97ed534b567270/integration_tests__asan__analyzer__[1_6].html
- https://s3.amazonaws.com/clickhouse-test-reports/60497/cb8d8594533de3773ad1f30f7a5792f1e7e1bb39/integration_tests__tsan__[1_6].html

which is super sporadic given how many CI tests run per day.

In all cases, we ran three queries with query cache TTL = 1 sec, then wait 2 sec, and find that the query cache contains only 1 entry whereas 2 entries are expected. I can honestly not think of a scenario where this can happen except one: the *asynchronous* metrics are updated in a weird time interval that they report only 1 entry. Instead of adding retries (l. 105, please revert all retries from this file which only hide problems), please add `SYSTEM RELOAD ASYNCHRONOUS METRICS` before l. 106, see https://github.com/ClickHouse/ClickHouse/pull/53710. When I am back from vacation, I can add some documentation for SYSTEM RELOAD ASYNCHRONOUS METRICS.

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #61869 from fm4v/ClickHouse:fix-flaky-tests2 -->
*This is an automated comment for commit ceaae464f914f269a02f4b0a168d7fdca42d1508 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61869/ceaae464f914f269a02f4b0a168d7fdca42d1508/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

fm4v committed:
integration tests: persist pdb history between test runs

Run pytest --pdb. On failure or breakpoint, it will fall back to PDB. Commands written here will be saved to a local file and loaded at the start of the next test run.

fm4v committed:
integration tests: add test `test_replicated_database_compare_parts`

fm4v committed:
stateless tests: fix linter

rschu1ze commented on tests/integration/test_runtime_configurable_cache_size/test.py:
```diff
@@ -101,9 +101,11 @@ def test_query_cache_size_is_runtime_configurable(start_cluster):
+    time.sleep(2)
     res = node.query_with_retry(
-        "SELECT value FROM system.asynchronous_metrics WHERE metric = 'QueryCacheEntries'",
+        "SELECT value FROM system.asynchronous_metrics WHERE metric = 'QueryCacheEntries' SETTINGS use_query_cache=0",
```
`use_query_cache = 0` is the default anyways, this has no effect, please remove.

azat commented on tests/integration/test_dictionaries_update_and_reload/test.py:
```diff
@@ -250,6 +260,15 @@ def test_reload_after_fail_by_timer(started_cluster):
+    # add retry for these builds
+    if (
+        instance.is_built_with_sanitizer()
+        and get_status("no_file_2") == "FAILED_AND_RELOADING"
```
Why it happens?

And it still fails - https://s3.amazonaws.com/clickhouse-test-reports/66311/316dd790bdb5a8cb4aef11c356f14700f3b9169e/integration_tests__asan__old_analyzer__[3_6].html

azat replied:
Anyway this code is wrong, since you still can get `FAILED_AND_RELOADING` two lines below, there should be `else` at least.

azat replied:
- https://github.com/ClickHouse/ClickHouse/pull/67754

fm4v replied:
https://github.com/ClickHouse/ClickHouse/pull/67754#issuecomment-2267598185
